target: introduce default location for external target linux generic

Ansuel opened the pull request:
An external target provided by a custom feed might depend on a specific
kernel version and might need to apply generic patches for the specific
kernel version.

Now that the kernel version file is defined in the generic target
director, it's possible to handle this scenario for an external target
from feed.

To do this, we introduce additional logic for GENERIC_PLATFORM_DIR where
we check if the target feed have in his upper directory a linux/generic
path and use that instead of the default TOPLEVEL linux/generic path.

An example for an external target from feed would have the following
path structure:

mediatek/Makefile
linux/generic
linux/generic/config-6.6
...

An external target from feed is also defined by creating a symbolic link
in target/linux/${BOARD} to the feed directory with board as the name of
the external target.

This is to try to enforce a more consistent way to handle external
target in downstream project that makes use of OpenWrt build system.

---

To add more info to this, in also every downstream SDK stuff is included to
OpenWrt with this hack of providing stuff with feeds. Introducing these
kind of change at least improves from their side the situation by forcing
them to use a more organized repository structure.

Ansuel committed:
target: don't use GENERIC_PLATFORM_DIR for generic other-files

Generic other-files directory path hardly needs to be overwritten as 99%
of the time the mainline generic one is usable.

Don't depend on GENERIC_PLATFORM_DIR variable (that can point to the
external target feed path) but introduce a dedicated define and set it
to the default TOPLEVEL location by default.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
target: introduce default location for external target linux generic

An external target provided by a custom feed might depend on a specific
kernel version and might need to apply generic patches for the specific
kernel version.

Now that the kernel version file is defined in the generic target
director, it's possible to handle this scenario for an external target
from feed.

To do this, we introduce additional logic for GENERIC_PLATFORM_DIR where
we check if the target feed have in his upper directory a linux/generic
path and use that instead of the default TOPLEVEL linux/generic path.

An example for an external target from feed would have the following
path structure:

mediatek/Makefile
linux/generic
linux/generic/config-6.6
...

An external target from feed is also defined by creating a symbolic link
in target/linux/${BOARD} to the feed directory with board as the name of
the external target.

This is to try to enforce a more consistent way to handle external
target in downstream project that makes use of OpenWrt build system.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
We need to balance flexibility and good practice.

Permitting overwrite permits random placement of the directory and all kind
of funny path.

Enforcing it on a default directory path force good practice while still
permitting to provide an external feed.

Il Lun 26 Mag 2025, 18:49 Andreas Gnau ***@***.***> ha
scritto:

> *Rondom* left a comment (openwrt/openwrt#18820)
> <https://github.com/openwrt/openwrt/pull/18820#issuecomment-2910258899>
>
> It is quite clever to leverage relative paths to achieve the desired
> result here. Minmal change, maximum usefullness.
>
> I am wondering if it is maybe "less magic" and more flexible if
> feed-targets could override the path to the generic directories from within
> the Makefile. That would make things explicit.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/18820#issuecomment-2910258899>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQTAH32F3IG4UWO2RNT3ANA2JAVCNFSM6AAAAAB5JRPZBOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMJQGI2TQOBZHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Rondom reviewed (approved):
Just tried this. It's great and makes things much simpler. I agree also that it is better to encourage/enforce the same structure.

KanjiMonster commented:
Does this also apply to `toolchain/kernel-headers`? The custom target's generic kernel may include uapi patches/additions not included in the default kernel, and without also changing kernel-headers these will be unavailable for user space.

Rondom commented:
> @KanjiMonster I don't think we should include target specific uapi patches for the toolchain build. If specific user space components depend on modified kernel includes, they can add the kernel headers generated early in target/compile to the include path.

I agree. Including target-specific headers in toolchain build would also mean that you would need to run `make targetclean` between builds if the target changes, because you do not now what kind of shenanigans each target did with the uapi.

Rondom commented:
... It also follows the existing design that we have right now, where the toolchain build does not include target-specific patches (most likely for exactly this reason.

Rondom commented:
> And we already have a situation where the selected target has an influence on the kernel headers, as IIRC the selected kernel version defines where the kernel headers are taken from.

Upstream kernel community cares about not breaking ABI and is way more trustworthy not do bad things than $random_silicon_vendor. I see your point, though.

In general, I think it should be easy to do things right and it should be hard to do things "wrong", and if the latter is required it should have minimum impact.
I certainly don't like that you would need to build different toolchains just because of one patch to one uapi header for example and I do not like my build breaking because of switching targets. The only clean way to fix the negative effects of having target-specific uapi-patches would be to compile one toolchain per target, which is wasteful. @nbd168's suggestion to override things selectively in `target/compile` sounds like an alternative solution without these downsides in my opinion.

Ansuel committed:
target: use SCAN_DIR when generating index for external feed

On generating the index for an external feed, use the SCAN_DIR path as
the target still needs to be installed (with scripts/feeds install ...)
and the link in target/linux/feeds created.

This is to handle corner case where an external feed provide a generic
linux version not present in TOPDIR target/linux. In such case
kernel-version.mk will complain of not finding the related kernel detail
file as the external feed target link is created only when the feed is
actually installed.

To permit correct creation of external feed index, only for the scan
phase, use the SCAN_DIR path that point to the feeds/ directory instead
of the target/linux/feeds one.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
